# **Capstone III Data Wrangling**

### **Section 1: Imports and Initial Inspection**

In [1]:
import pandas as pd

In [2]:
train = pd.read_csv("../Raw Data/train.csv")
test = pd.read_csv('../Raw Data/test.csv')

In [3]:
train.head()

Id        Date   Time  CO2_(dinning-room)  CO2_room  \
0   0  13/03/2012  11:45             216.560   221.920   
1   1  13/03/2012  12:00             219.947   220.363   
2   2  13/03/2012  12:15             219.403   218.933   
3   3  13/03/2012  12:30             218.613   217.045   
4   4  13/03/2012  12:45             217.714   216.080   

   Relative_humidity_(dinning-room)  Relative_humidity_room  \
0                           39.9125                 42.4150   
1                           39.9267                 42.2453   
2                           39.7720                 42.2267   
3                           39.7760                 42.0987   
4                           39.7757                 42.0686   

   Lighting_(dinning-room)  Lighting_room  Meteo_Rain  Meteo_Sun_dusk  \
0                  81.6650        113.520         0.0         623.360   
1                  81.7413        113.605         0.0         623.211   
2                  81.4240        113.600         0.0         622.656   
3                  81.5013        113.344         0.0         622.571   
4                  81.4657        113.034         0.0         622.400   

   Meteo_Wind  Meteo_Sun_light_in_west_facade  Meteo_Sun_light_in_east_facade  \
0     1.42625                         9690.24                        12604.20   
1     1.59200                        11022.00                        10787.20   
2     1.89133                        13960.50                         9669.63   
3     1.82800                        18511.20                         9648.13   
4     2.36071                        26349.00                         9208.32   

   Meteo_Sun_light_in_south_facade  Meteo_Sun_irradiance  \
0                          95436.8               758.880   
1                          95436.8               762.069   
2                          95398.6               766.251   
3                          95360.3               766.037   
4                          95354.9               762.743   

   Outdoor_relative_humidity_Sensor  Day_of_the_week  Indoor_temperature_room  
0                           48.3750              2.0                  17.8275  
1                           47.8080              2.0                  18.1207  
2                           47.4320              2.0                  18.4367  
3                           47.0240              2.0                  18.7513  
4                           45.4743              2.0                  19.0414

In [4]:
test.head()

Id        Date   Time  CO2_(dinning-room)  CO2_room  \
0  2764  18/04/2012  00:00             207.840   206.560   
1  2765  18/04/2012  00:15             207.499   206.101   
2  2766  18/04/2012  00:30             207.360   206.805   
3  2767  18/04/2012  00:45             207.445   206.325   
4  2768  18/04/2012  01:00             207.840   205.952   

   Relative_humidity_(dinning-room)  Relative_humidity_room  \
0                           31.5800                 30.1600   
1                           31.6240                 30.0613   
2                           31.6507                 30.1773   
3                           31.6987                 30.2560   
4                           31.7707                 30.3840   

   Lighting_(dinning-room)  Lighting_room  Meteo_Rain  Meteo_Sun_dusk  \
0                  11.7200        14.1600         0.0            0.65   
1                  11.5240        14.7600         0.0            0.65   
2                  11.6220        14.9547         0.0            0.65   
3                  11.7847        15.0520         0.0            0.65   
4                  11.5887        15.2147         0.0            0.65   

   Meteo_Wind  Meteo_Sun_light_in_west_facade  Meteo_Sun_light_in_east_facade  \
0    0.000000                             0.0                             0.0   
1    0.000000                             0.0                             0.0   
2    0.088000                             0.0                             0.0   
3    0.127333                             0.0                             0.0   
4    0.000000                             0.0                             0.0   

   Meteo_Sun_light_in_south_facade  Meteo_Sun_irradiance  \
0                              0.0              -3.12000   
1                              0.0              -2.80800   
2                              0.0              -2.96400   
3                              0.0              -3.02933   
4                              0.0              -2.88600   

   Outdoor_relative_humidity_Sensor  Day_of_the_week  
0                           39.0800              3.0  
1                           38.9680              3.0  
2                           38.9467              3.0  
3                           39.0760              3.0  
4                           40.0413              3.0

In [5]:
train.shape

(2764, 19)

In [6]:
test.shape

(1373, 18)

In [7]:
# With the exception of Date and Time, all columns are the correct data type
train.dtypes

Id                                    int64
Date                                 object
Time                                 object
CO2_(dinning-room)                  float64
CO2_room                            float64
Relative_humidity_(dinning-room)    float64
Relative_humidity_room              float64
Lighting_(dinning-room)             float64
Lighting_room                       float64
Meteo_Rain                          float64
Meteo_Sun_dusk                      float64
Meteo_Wind                          float64
Meteo_Sun_light_in_west_facade      float64
Meteo_Sun_light_in_east_facade      float64
Meteo_Sun_light_in_south_facade     float64
Meteo_Sun_irradiance                float64
Outdoor_relative_humidity_Sensor    float64
Day_of_the_week                     float64
Indoor_temperature_room             float64
dtype: object

In [8]:
test.dtypes

Id                                    int64
Date                                 object
Time                                 object
CO2_(dinning-room)                  float64
CO2_room                            float64
Relative_humidity_(dinning-room)    float64
Relative_humidity_room              float64
Lighting_(dinning-room)             float64
Lighting_room                       float64
Meteo_Rain                          float64
Meteo_Sun_dusk                      float64
Meteo_Wind                          float64
Meteo_Sun_light_in_west_facade      float64
Meteo_Sun_light_in_east_facade      float64
Meteo_Sun_light_in_south_facade     float64
Meteo_Sun_irradiance                float64
Outdoor_relative_humidity_Sensor    float64
Day_of_the_week                     float64
dtype: object

In [9]:
# No NaN values
train.isna().sum()

Id                                  0
Date                                0
Time                                0
CO2_(dinning-room)                  0
CO2_room                            0
Relative_humidity_(dinning-room)    0
Relative_humidity_room              0
Lighting_(dinning-room)             0
Lighting_room                       0
Meteo_Rain                          0
Meteo_Sun_dusk                      0
Meteo_Wind                          0
Meteo_Sun_light_in_west_facade      0
Meteo_Sun_light_in_east_facade      0
Meteo_Sun_light_in_south_facade     0
Meteo_Sun_irradiance                0
Outdoor_relative_humidity_Sensor    0
Day_of_the_week                     0
Indoor_temperature_room             0
dtype: int64

In [10]:
test.isna().sum()

Id                                  0
Date                                0
Time                                0
CO2_(dinning-room)                  0
CO2_room                            0
Relative_humidity_(dinning-room)    0
Relative_humidity_room              0
Lighting_(dinning-room)             0
Lighting_room                       0
Meteo_Rain                          0
Meteo_Sun_dusk                      0
Meteo_Wind                          0
Meteo_Sun_light_in_west_facade      0
Meteo_Sun_light_in_east_facade      0
Meteo_Sun_light_in_south_facade     0
Meteo_Sun_irradiance                0
Outdoor_relative_humidity_Sensor    0
Day_of_the_week                     0
dtype: int64

### **Section 2: Format and Structure Changes**

Here we'll make some format changes. This includes creating a combined date and time column, changing the data types of the date columns, and changing column names.

In [11]:
train.head()

Id        Date   Time  CO2_(dinning-room)  CO2_room  \
0   0  13/03/2012  11:45             216.560   221.920   
1   1  13/03/2012  12:00             219.947   220.363   
2   2  13/03/2012  12:15             219.403   218.933   
3   3  13/03/2012  12:30             218.613   217.045   
4   4  13/03/2012  12:45             217.714   216.080   

   Relative_humidity_(dinning-room)  Relative_humidity_room  \
0                           39.9125                 42.4150   
1                           39.9267                 42.2453   
2                           39.7720                 42.2267   
3                           39.7760                 42.0987   
4                           39.7757                 42.0686   

   Lighting_(dinning-room)  Lighting_room  Meteo_Rain  Meteo_Sun_dusk  \
0                  81.6650        113.520         0.0         623.360   
1                  81.7413        113.605         0.0         623.211   
2                  81.4240        113.600         0.0         622.656   
3                  81.5013        113.344         0.0         622.571   
4                  81.4657        113.034         0.0         622.400   

   Meteo_Wind  Meteo_Sun_light_in_west_facade  Meteo_Sun_light_in_east_facade  \
0     1.42625                         9690.24                        12604.20   
1     1.59200                        11022.00                        10787.20   
2     1.89133                        13960.50                         9669.63   
3     1.82800                        18511.20                         9648.13   
4     2.36071                        26349.00                         9208.32   

   Meteo_Sun_light_in_south_facade  Meteo_Sun_irradiance  \
0                          95436.8               758.880   
1                          95436.8               762.069   
2                          95398.6               766.251   
3                          95360.3               766.037   
4                          95354.9               762.743   

   Outdoor_relative_humidity_Sensor  Day_of_the_week  Indoor_temperature_room  
0                           48.3750              2.0                  17.8275  
1                           47.8080              2.0                  18.1207  
2                           47.4320              2.0                  18.4367  
3                           47.0240              2.0                  18.7513  
4                           45.4743              2.0                  19.0414

In [12]:
test.head()

Id        Date   Time  CO2_(dinning-room)  CO2_room  \
0  2764  18/04/2012  00:00             207.840   206.560   
1  2765  18/04/2012  00:15             207.499   206.101   
2  2766  18/04/2012  00:30             207.360   206.805   
3  2767  18/04/2012  00:45             207.445   206.325   
4  2768  18/04/2012  01:00             207.840   205.952   

   Relative_humidity_(dinning-room)  Relative_humidity_room  \
0                           31.5800                 30.1600   
1                           31.6240                 30.0613   
2                           31.6507                 30.1773   
3                           31.6987                 30.2560   
4                           31.7707                 30.3840   

   Lighting_(dinning-room)  Lighting_room  Meteo_Rain  Meteo_Sun_dusk  \
0                  11.7200        14.1600         0.0            0.65   
1                  11.5240        14.7600         0.0            0.65   
2                  11.6220        14.9547         0.0            0.65   
3                  11.7847        15.0520         0.0            0.65   
4                  11.5887        15.2147         0.0            0.65   

   Meteo_Wind  Meteo_Sun_light_in_west_facade  Meteo_Sun_light_in_east_facade  \
0    0.000000                             0.0                             0.0   
1    0.000000                             0.0                             0.0   
2    0.088000                             0.0                             0.0   
3    0.127333                             0.0                             0.0   
4    0.000000                             0.0                             0.0   

   Meteo_Sun_light_in_south_facade  Meteo_Sun_irradiance  \
0                              0.0              -3.12000   
1                              0.0              -2.80800   
2                              0.0              -2.96400   
3                              0.0              -3.02933   
4                              0.0              -2.88600   

   Outdoor_relative_humidity_Sensor  Day_of_the_week  
0                           39.0800              3.0  
1                           38.9680              3.0  
2                           38.9467              3.0  
3                           39.0760              3.0  
4                           40.0413              3.0

In [13]:
# Drop the Id columns. We'll be indexing our dataframe by date/time.
train.drop(columns = 'Id', inplace = True)
test.drop(columns = 'Id', inplace = True)

In [14]:
# We'd like these column names to be formatted differently.
train.columns

Index(['Date', 'Time', 'CO2_(dinning-room)', 'CO2_room',
       'Relative_humidity_(dinning-room)', 'Relative_humidity_room',
       'Lighting_(dinning-room)', 'Lighting_room', 'Meteo_Rain',
       'Meteo_Sun_dusk', 'Meteo_Wind', 'Meteo_Sun_light_in_west_facade',
       'Meteo_Sun_light_in_east_facade', 'Meteo_Sun_light_in_south_facade',
       'Meteo_Sun_irradiance', 'Outdoor_relative_humidity_Sensor',
       'Day_of_the_week', 'Indoor_temperature_room'],
      dtype='object')

In [15]:
test.columns

Index(['Date', 'Time', 'CO2_(dinning-room)', 'CO2_room',
       'Relative_humidity_(dinning-room)', 'Relative_humidity_room',
       'Lighting_(dinning-room)', 'Lighting_room', 'Meteo_Rain',
       'Meteo_Sun_dusk', 'Meteo_Wind', 'Meteo_Sun_light_in_west_facade',
       'Meteo_Sun_light_in_east_facade', 'Meteo_Sun_light_in_south_facade',
       'Meteo_Sun_irradiance', 'Outdoor_relative_humidity_Sensor',
       'Day_of_the_week'],
      dtype='object')

In [16]:
# Function to remove parentheses, fix mispellings, make lower case, and change dashes to underscores in a list of strings.
def column_renamer(df):
    new_names = []
    for column in df.columns:
        new_name = column.lower()
        new_name = (new_name.replace('(', '').replace(')', '').replace('-', '_')
        .replace('dinning', 'dining').replace('meteo_', '').replace('_sensor', ''))
        new_names.append(new_name)
    return(new_names)

In [17]:
# Replace column names with the formatted names using our column_renamer function.
new_columns_train = column_renamer(train)
new_columns_test = column_renamer(test)

mapper_train = dict(zip(train.columns, new_columns_train))
mapper_test = dict(zip(test.columns, new_columns_test))

train.rename(mapper_train, axis = 'columns', inplace = True)
test.rename(mapper_test, axis = 'columns', inplace = True)
train.head()

date   time  co2_dining_room  co2_room  \
0  13/03/2012  11:45          216.560   221.920   
1  13/03/2012  12:00          219.947   220.363   
2  13/03/2012  12:15          219.403   218.933   
3  13/03/2012  12:30          218.613   217.045   
4  13/03/2012  12:45          217.714   216.080   

   relative_humidity_dining_room  relative_humidity_room  \
0                        39.9125                 42.4150   
1                        39.9267                 42.2453   
2                        39.7720                 42.2267   
3                        39.7760                 42.0987   
4                        39.7757                 42.0686   

   lighting_dining_room  lighting_room  rain  sun_dusk     wind  \
0               81.6650        113.520   0.0   623.360  1.42625   
1               81.7413        113.605   0.0   623.211  1.59200   
2               81.4240        113.600   0.0   622.656  1.89133   
3               81.5013        113.344   0.0   622.571  1.82800   
4               81.4657        113.034   0.0   622.400  2.36071   

   sun_light_in_west_facade  sun_light_in_east_facade  \
0                   9690.24                  12604.20   
1                  11022.00                  10787.20   
2                  13960.50                   9669.63   
3                  18511.20                   9648.13   
4                  26349.00                   9208.32   

   sun_light_in_south_facade  sun_irradiance  outdoor_relative_humidity  \
0                    95436.8         758.880                    48.3750   
1                    95436.8         762.069                    47.8080   
2                    95398.6         766.251                    47.4320   
3                    95360.3         766.037                    47.0240   
4                    95354.9         762.743                    45.4743   

   day_of_the_week  indoor_temperature_room  
0              2.0                  17.8275  
1              2.0                  18.1207  
2              2.0                  18.4367  
3              2.0                  18.7513  
4              2.0                  19.0414

In [18]:
test.head()

date   time  co2_dining_room  co2_room  \
0  18/04/2012  00:00          207.840   206.560   
1  18/04/2012  00:15          207.499   206.101   
2  18/04/2012  00:30          207.360   206.805   
3  18/04/2012  00:45          207.445   206.325   
4  18/04/2012  01:00          207.840   205.952   

   relative_humidity_dining_room  relative_humidity_room  \
0                        31.5800                 30.1600   
1                        31.6240                 30.0613   
2                        31.6507                 30.1773   
3                        31.6987                 30.2560   
4                        31.7707                 30.3840   

   lighting_dining_room  lighting_room  rain  sun_dusk      wind  \
0               11.7200        14.1600   0.0      0.65  0.000000   
1               11.5240        14.7600   0.0      0.65  0.000000   
2               11.6220        14.9547   0.0      0.65  0.088000   
3               11.7847        15.0520   0.0      0.65  0.127333   
4               11.5887        15.2147   0.0      0.65  0.000000   

   sun_light_in_west_facade  sun_light_in_east_facade  \
0                       0.0                       0.0   
1                       0.0                       0.0   
2                       0.0                       0.0   
3                       0.0                       0.0   
4                       0.0                       0.0   

   sun_light_in_south_facade  sun_irradiance  outdoor_relative_humidity  \
0                        0.0        -3.12000                    39.0800   
1                        0.0        -2.80800                    38.9680   
2                        0.0        -2.96400                    38.9467   
3                        0.0        -3.02933                    39.0760   
4                        0.0        -2.88600                    40.0413   

   day_of_the_week  
0              3.0  
1              3.0  
2              3.0  
3              3.0  
4              3.0

In [19]:
# Add a new column that combines the Date and Time columns. Drop the time column.
train['date_time'] = train['date'] + ' ' + train['time']
train.drop(columns = 'time', inplace = True)

test['date_time'] = test['date'] + ' ' + test['time']
test.drop(columns = 'time', inplace = True)

train['date_time'].sample(10)

1863    01/04/2012 21:30
2584    09/04/2012 09:45
2428    07/04/2012 18:45
2661    10/04/2012 05:00
2116    04/04/2012 12:45
758     21/03/2012 09:15
1915    02/04/2012 10:30
710     20/03/2012 21:15
536     19/03/2012 01:45
2557    09/04/2012 03:00
Name: date_time, dtype: object

In [20]:
test['date_time'].sample(10)

27      18/04/2012 06:45
816     26/04/2012 12:00
752     25/04/2012 20:00
403     22/04/2012 04:45
1157    30/04/2012 01:15
730     25/04/2012 14:30
236     20/04/2012 11:00
859     26/04/2012 22:45
737     25/04/2012 16:15
400     22/04/2012 04:00
Name: date_time, dtype: object

In [21]:
# These should be datetime objects.
train[['date', 'date_time']].dtypes, test[['date', 'date_time']].dtypes

(date         object
 date_time    object
 dtype: object,
 date         object
 date_time    object
 dtype: object)

In [22]:
# Change Date, Time, and Date_time columns to be datetime objects instead of strings.
train['date'] = pd.to_datetime(train['date'], format = '%d/%m/%Y')
train['date_time'] = pd.to_datetime(train['date_time'], format = "%d/%m/%Y %H:%M")

test['date'] = pd.to_datetime(test['date'], format = '%d/%m/%Y')
test['date_time'] = pd.to_datetime(test['date_time'], format = "%d/%m/%Y %H:%M")

train[['date', 'date_time']].sample(10)

date           date_time
2647 2012-04-10 2012-04-10 01:30:00
144  2012-03-14 2012-03-14 23:45:00
1018 2012-03-24 2012-03-24 02:15:00
2747 2012-04-11 2012-04-11 02:30:00
1961 2012-04-02 2012-04-02 22:00:00
751  2012-03-21 2012-03-21 07:30:00
1194 2012-03-25 2012-03-25 22:15:00
584  2012-03-19 2012-03-19 13:45:00
77   2012-03-14 2012-03-14 07:00:00
2023 2012-04-03 2012-04-03 13:30:00

In [23]:
test[['date', 'date_time']].sample(10)

date           date_time
391  2012-04-22 2012-04-22 01:45:00
171  2012-04-19 2012-04-19 18:45:00
1335 2012-05-01 2012-05-01 21:45:00
1248 2012-05-01 2012-05-01 00:00:00
454  2012-04-22 2012-04-22 17:30:00
682  2012-04-25 2012-04-25 02:30:00
34   2012-04-18 2012-04-18 08:30:00
1271 2012-05-01 2012-05-01 05:45:00
456  2012-04-22 2012-04-22 18:00:00
707  2012-04-25 2012-04-25 08:45:00

In [24]:
train.dtypes, test.dtypes

(date                             datetime64[ns]
 co2_dining_room                         float64
 co2_room                                float64
 relative_humidity_dining_room           float64
 relative_humidity_room                  float64
 lighting_dining_room                    float64
 lighting_room                           float64
 rain                                    float64
 sun_dusk                                float64
 wind                                    float64
 sun_light_in_west_facade                float64
 sun_light_in_east_facade                float64
 sun_light_in_south_facade               float64
 sun_irradiance                          float64
 outdoor_relative_humidity               float64
 day_of_the_week                         float64
 indoor_temperature_room                 float64
 date_time                        datetime64[ns]
 dtype: object,
 date                             datetime64[ns]
 co2_dining_room                         float64
 co2

### **Section 3: Data Validation**

Now we'll inspect the contents of our columns to make sure the data makes sense. We know there are no missing values, but we need to make sure the data isn't invalid in other ways. For example, we shouldn't see negative values in columns like the co2 columns, which are in parts per million. We also need to see if any measurements were skipped. They should occur every fifteen minutes.

In [25]:
# Measurements are supposed to have been taken every 15 minutes. Let's find out if any rows are missing.
# .diff() takes the difference between each row and the previous row.
train['date_time'].diff().drop([0]).max(), train['date_time'].diff().drop([0]).min()

(Timedelta('0 days 00:15:00'), Timedelta('0 days 00:15:00'))

In [26]:
test['date_time'].diff().drop([0]).max(), test['date_time'].diff().drop([0]).min()

(Timedelta('0 days 00:30:00'), Timedelta('0 days 00:15:00'))

Train is fine, meaning the maximum and minimum difference in time between any two subsequent measurments is 15 minutes. In the test set it looks like there is a maximum difference of 30 minutes, meaning at least one measurement was likely skipped. We must investigate this.

In [30]:
# There are two instances of a gap of 30 minutes between measurements
test['date_time'].diff().drop([0]).sort_values(ascending = False)

1362   0 days 00:30:00
1356   0 days 00:30:00
912    0 days 00:15:00
920    0 days 00:15:00
919    0 days 00:15:00
             ...      
456    0 days 00:15:00
455    0 days 00:15:00
454    0 days 00:15:00
453    0 days 00:15:00
1372   0 days 00:15:00
Name: date_time, Length: 1372, dtype: timedelta64[ns]

In [34]:
# Our culprits both occur on May 2, from 02:45 to 3:15, and again from 4:30 to 5:00.
test['date_time'].iloc[1355:1363]

1355   2012-05-02 02:45:00
1356   2012-05-02 03:15:00
1357   2012-05-02 03:30:00
1358   2012-05-02 03:45:00
1359   2012-05-02 04:00:00
1360   2012-05-02 04:15:00
1361   2012-05-02 04:30:00
1362   2012-05-02 05:00:00
Name: date_time, dtype: datetime64[ns]

There are different strategies for imputing missing data. In non time series data sets, common strategies include imputing the mean or median values for the missing values. For a time series data set like ours, strategies include duplicating the last value, duplicating the next value, linear interpolation, and 'spline' interpolation, where the goal is to minimize the curvature of the imputed data point. Another strategy common to all data sets is to simply drop the missing values. In our current situation, we're only missing 2 out of 1373 data points. Our intutions about weather and temperature lead us to believe that it's very unlikely to be suprised by radical deviations in fifteen minute or half hour intervals (in other words, our data has high autocorrelation which we'll see in the EDA section). Therefore, we'd be inclined to drop the missing values which is exactly what the data uploaders have already done for us. We will not attempt to impute any values.

In [27]:
# Function to check if the minimum values of the numeric data types are above zero
def min_max_checker(df):
    invalid_value_columns = []
    for column in df.columns:
        try:
            df[column][0] + 1 # Checks to see if column is of numeric type
            if df[column].min() < 0: # If it is numeric, and if the minimum value is below zero, add it to the list
                invalid_value_columns.append(column)
        except:
            pass # Does nothing if the column is not numeric
    return(invalid_value_columns) # Return a list of columns with minimum values below zero.

In [28]:
min_max_checker(train), min_max_checker(test)

(['sun_irradiance'], ['sun_irradiance'])

In [29]:
train['sun_irradiance'].describe()

count    2764.000000
mean      215.010017
std       297.234046
min        -4.164670
25%        -3.381330
50%         3.922000
75%       435.434500
max      1028.270000
Name: sun_irradiance, dtype: float64

In [30]:
test['sun_irradiance'].describe()

count    1373.000000
mean      266.816322
std       338.532417
min        -3.708000
25%        -3.042000
50%        35.816000
75%       566.293000
max      1094.660000
Name: sun_irradiance, dtype: float64

In [31]:
# At least 25% of irradiance values are negative. Let's see exactly what proportion of irradiance's values are negative.
negative_irrad_rows = len(train[train['sun_irradiance'] < 0])
all_rows = len(train)
print(round(negative_irrad_rows / all_rows * 100, 2), 'percent of the training irradiance readings are below zero.')

49.13 percent of the training irradiance readings are below zero.


In [32]:
negative_irrad_rows = len(test[test['sun_irradiance'] < 0])
all_rows = len(test)
print(round(negative_irrad_rows / all_rows * 100, 2), 'percent of the testing irradiance readings are below zero.')

44.28 percent of the testing irradiance readings are below zero.


Upon cursory investigation, it seems to be relatively common to have small negative irradiance values. The reason for this might be because the value is recorded on a log scale, with an arbitrary zero point at, say, dawn or dusk. At night (which comprises half of the measurments), the absence of sun falls just below this zero irradiance point, meaning the value of the exponent variable is negative. During the day, it is vastly higher.

Posts that lead to this conclusion:

https://unmethours.com/question/13731/negative-and-positive-values-of-solar-radiation-during-night/
https://www.researchgate.net/post/Its_normal_to_have_negative_values_of_irradiance_in_a_database

In [33]:
# Lastly, we'll set the indices of the train and test sets to be our date_time columns.
train.set_index('date_time', inplace = True)
test.set_index('date_time', inplace = True)

In [34]:
# Save cleaned version of the data.
train.to_csv('../Data/train_cleaned.csv')
test.to_csv('../Data/test_cleaned.csv')

# Summary

Here is a summary of the changes we have made to the data so far:

1) Made combined date_time columns out of the date and time columns,

2) Changed the data types of the date and date_time columns to be datetime objects,

3) Deleted the time and Id columns,

4) Set the indices of our dataframes to be the combined date_time columns,

5) Fixed spelling errors and in general made the column name formatting consistent and less verbose.


The data was quite clean when it came to us. There were no missing values in the form of NaN's we needed to deal with. We further confirmed that the columns weren't missing values in other ways by observing the maximum and minimum values of the differences between each row. Train had no more and no less than a 15 minute difference between two subsequent rows, and test only had two instances 30 minutes between two subsequent rows (two missing values). These were safe to leave be. Lastly, we made sure that no column had negative values that wasn't supposed to have negative values. 